File: lessons/09-ollama-function-calling/python/ollama_chat.py
"""Lesson 9 - Ollama's /api/chat with tools, and nothing else.

Lesson 1's provider (`localrag/providers/ollama.py`) sends one system and one
user message and reads back text. Tool calling needs the whole message list and
a `tools` array, so this is a separate ~80-line client over `requests`, which the
course already depends on. No SDK: the point is to see the JSON.

Request:   {"model", "messages", "tools": [...], "stream": false, "options": {...}}
Response:  {"message": {"role": "assistant", "content": "...",
                        "tool_calls": [{"function": {"name": ..., "arguments": {...}}}]}}
"""

from __future__ import annotations

import time
from typing import Dict, List, Optional

import requests


class OllamaError(RuntimeError):
    """The model failed: a crash, a bad request, a missing model."""


class OllamaUnreachable(OllamaError):
    """The server is not there. Not the model's fault, so never recorded as its result."""


class OllamaModel:
    def __init__(self, url: str, model: str, *, think: Optional[bool] = False,
                 temperature: float = 0.0, seed: int = 7, num_ctx: int = 8192,
                 timeout: float = 600.0) -> None:
        self.url = url.rstrip("/")
        self.model = model
        self.think = think
        # num_ctx: Ollama's default window is 4K on machines with under 24 GB of VRAM.
        # Tool schemas, three passages and a few turns fit in 8K; past the window,
        # Ollama truncates the prompt and the model never sees what was cut.
        self.options = {"temperature": temperature, "seed": seed, "num_ctx": num_ctx}
        self.timeout = timeout
        self._caps: Optional[List[str]] = None

    def capabilities(self) -> List[str]:
        """What the model's own metadata claims: e.g. ['completion', 'tools', 'thinking']."""
        if self._caps is None:
            self._caps = show(self.url, self.model).get("capabilities", [])
        return self._caps

    def chat(self, messages: List[dict], tools: List[dict]) -> Dict:
        payload: Dict = {"model": self.model, "messages": messages, "stream": False,
                         "options": self.options}
        if tools:
            payload["tools"] = tools
        # Sending `think` to a model without the capability is an error, not a no-op.
        if self.think is not None and "thinking" in self.capabilities():
            payload["think"] = self.think
        started = time.monotonic()
        try:
            resp = requests.post(f"{self.url}/api/chat", json=payload, timeout=self.timeout)
        except requests.ConnectionError as exc:
            # First: ConnectTimeout is both a ConnectionError and a Timeout, and a
            # server that cannot be reached is an outage, not the model's answer.
            raise OllamaUnreachable(f"cannot reach Ollama at {self.url}: {exc}") from exc
        except requests.Timeout as exc:
            # The server is there; the model did not answer in time. On a CPU that
            # is a result worth recording, not an outage.
            raise OllamaError(f"no reply within {self.timeout:.0f}s") from exc
        except requests.RequestException as exc:
            raise OllamaUnreachable(f"cannot reach Ollama at {self.url}: {exc}") from exc
        if resp.status_code == 404:
            raise OllamaError(f"model {self.model!r} not found - run: ollama pull {self.model}")
        body = _json(resp)
        if resp.status_code != 200 or "error" in body:
            raise OllamaError(body.get("error") or f"HTTP {resp.status_code}")
        msg = body.get("message", {})
        keep = {"role": "assistant", "content": msg.get("content") or ""}
        if msg.get("tool_calls"):
            keep["tool_calls"] = msg["tool_calls"]
        return {"message": keep, "seconds": round(time.monotonic() - started, 1)}


def _json(resp) -> Dict:
    try:
        return resp.json()
    except ValueError:
        return {"error": resp.text[:200] or f"HTTP {resp.status_code}"}


def show(url: str, model: str) -> Dict:
    try:
        resp = requests.post(f"{url.rstrip('/')}/api/show", json={"model": model}, timeout=30)
    except requests.RequestException as exc:
        raise OllamaUnreachable(f"cannot reach Ollama at {url}: {exc}") from exc
    return _json(resp) if resp.status_code == 200 else {}


def unload(url: str, model: str) -> None:
    """Free the model's memory now instead of after Ollama's 5-minute keep-alive.

    Recording six models back to back on a 19 GB laptop got the server
    OOM-killed: the last model was still resident while the next one loaded.
    """
    try:
        requests.post(f"{url.rstrip('/')}/api/generate",
                      json={"model": model, "keep_alive": 0}, timeout=60)
    except requests.RequestException:
        pass


def version(url: str) -> str:
    try:
        return requests.get(f"{url.rstrip('/')}/api/version", timeout=5).json().get("version", "?")
    except (requests.RequestException, ValueError):
        return "?"


def local_models(url: str) -> List[Dict]:
    """Every pulled model with its size and advertised capabilities."""
    try:
        resp = requests.get(f"{url.rstrip('/')}/api/tags", timeout=10)
        resp.raise_for_status()
    except requests.RequestException as exc:
        raise OllamaUnreachable(f"cannot reach Ollama at {url}: {exc}") from exc
    out = []
    for m in resp.json().get("models", []):
        caps = show(url, m["name"]).get("capabilities", [])
        gb = round(m.get("size", 0) / 1e9, 1)
        out.append({"name": m["name"], "gb": gb, "capabilities": caps})
    return sorted(out, key=lambda m: m["name"])

File: lessons/09-ollama-function-calling/node/ollama_chat.mjs
// Lesson 9 - Ollama's /api/chat with tools, and nothing else.
//
// Mirrors python/ollama_chat.py over Node's built-in fetch: no SDK, no npm
// package, because the point is to see the JSON.
//
// Request:   {"model", "messages", "tools": [...], "stream": false, "options": {...}}
// Response:  {"message": {"role": "assistant", "content": "...",
//                         "tool_calls": [{"function": {"name": ..., "arguments": {...}}}]}}

import net from "node:net";

import { RuntimeError } from "./cassette.mjs";
import { dumps, formatFixed, isDict, loads, pyRound, repr } from "./pycompat.mjs";

/** The model failed: a crash, a bad request, a missing model. */
export class OllamaError extends RuntimeError {}
/** The server is not there. Not the model's fault. */
export class OllamaUnreachable extends OllamaError {}

// Parse with pycompat.loads so a float argument stays a float, as it does when
// Python's requests parses the same body.
async function readJson(resp) {
  const text = await resp.text();
  try {
    return loads(text);
  } catch {
    return { error: text.slice(0, 200) || `HTTP ${resp.status}` };
  }
}

// Could a TCP connection to the server be opened at all, within `ms`?
function reachable(base, ms = 5_000) {
  const url = new URL(base);
  const port = Number(url.port) || (url.protocol === "https:" ? 443 : 80);
  return new Promise((resolve) => {
    const sock = net.connect({ host: url.hostname, port, timeout: ms });
    const done = (ok) => {
      sock.destroy();
      resolve(ok);
    };
    sock.once("connect", () => done(true)).once("timeout", () => done(false)).once("error", () => done(false));
  });
}

async function post(base, route, body, timeoutMs) {
  try {
    return await fetch(`${base}${route}`, {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      // dumps, not JSON.stringify: a PyFloat argument must go back as 3.0, not {"value":3}.
      body: dumps(body),
      signal: AbortSignal.timeout(timeoutMs),
    });
  } catch (exc) {
    // fetch has one deadline for connect and reply together, where requests has
    // one for each. So when it fires, probe the port: a server that answers is
    // there and the model was slow (a result); one that does not is an outage.
    if (exc.name === "TimeoutError" && (await reachable(base))) {
      throw new OllamaError(`no reply within ${formatFixed(timeoutMs / 1000, 0)}s`);
    }
    throw new OllamaUnreachable(`cannot reach Ollama at ${base}: ${exc.cause?.message ?? exc.message}`);
  }
}

export async function show(url, model) {
  const resp = await post(url.replace(/\/+$/, ""), "/api/show", { model }, 30_000);
  const body = resp.status === 200 ? await readJson(resp) : {};
  return isDict(body) ? body : {};
}

export class OllamaModel {
  constructor(url, model, { think = false, temperature = 0.0, seed = 7, numCtx = 8192, timeoutMs = 600_000 } = {}) {
    this.url = url.replace(/\/+$/, "");
    this.model = model;
    this.think = think;
    // num_ctx: Ollama's default window is 4K on machines with under 24 GB of VRAM.
    // Tool schemas, three passages and a few turns fit in 8K; past the window,
    // Ollama truncates the prompt and the model never sees what was cut.
    this.options = { temperature, seed, num_ctx: numCtx };
    this.timeoutMs = timeoutMs;
    this.caps = null;
  }

  /** What the model's own metadata claims: e.g. ['completion', 'tools', 'thinking']. */
  async capabilities() {
    if (this.caps === null) this.caps = (await show(this.url, this.model)).capabilities ?? [];
    return this.caps;
  }

  async chat(messages, tools) {
    const payload = { model: this.model, messages, stream: false, options: this.options };
    if (tools.length) payload.tools = tools;
    // Sending `think` to a model without the capability is an error, not a no-op.
    if (this.think !== null && (await this.capabilities()).includes("thinking")) payload.think = this.think;
    const started = performance.now();
    const resp = await post(this.url, "/api/chat", payload, this.timeoutMs);
    if (resp.status === 404) {
      throw new OllamaError(`model ${repr(this.model)} not found - run: ollama pull ${this.model}`);
    }
    const parsed = await readJson(resp);
    const body = isDict(parsed) ? parsed : {};
    if (resp.status !== 200 || "error" in body) throw new OllamaError(body.error || `HTTP ${resp.status}`);
    const msg = isDict(body.message) ? body.message : {};
    const keep = { role: "assistant", content: msg.content || "" };
    if (Array.isArray(msg.tool_calls) && msg.tool_calls.length) keep.tool_calls = msg.tool_calls;
    return { message: keep, seconds: pyRound((performance.now() - started) / 1000, 1) };
  }
}
